### PR TITLE
MdeModulePkg/Bus/Pci/NvmExpressDxe: Nvm Express Media Sanitize Protocol. [Rebase & FF]

### PR DESCRIPTION
**Pull request by apop5**
## Description

Implementation of MEDIA_SANITIZE_PROTOCOL for NIST purge/clear actions with mapping to NVM Express native commands.

## Cherry-Pick and squash the following commits:
[3e9cf8cdbd](https://github.com/microsoft/mu_basecore/commit/3e9cf8cdbd)

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested
N/A

## Integration Instructions
N/A

### CODE REVIEW

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:1]: Same comment as in other PRs to preserve authors in RB&FF commits.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c [hunk 1]
```diff
@@ -2,7 +2,8 @@
   NvmExpressDxe driver is used to manage non-volatile memory subsystem which follows
   NVM Express specification.
 
-  Copyright (c) 2013 - 2017, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) 2013 - 2017, Intel Corporation. All rights reserved.
```
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c]: I don't think it was necessary to drop `<BR>` in Intel's copyright.

Comment (makubacki) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c]: Similar in other files. It also creates an extra delta from edk2.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h [hunk 2]
```diff
@@ -75,6 +78,45 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
...
+#define NVME_CQE_SCT_CMD_SPECIFIC_STATUS                 0x1
+#define NVME_CQE_SCT_MEDIA_DATA_INTEGRITY_ERRORS_STATUS  0x2
+#define NVME_CQE_SCT_PATH_RELATED_STATUS                 0x3
+#define NVME_CQE_SC_SUCCESSFUL_COMPLETION                0x00
+#define NVME_CQE_SC_INVALID_CMD_OPCODE                   0x01
+#define NVME_CQE_SC_INVALID_FIELD_IN_CMD                 0x02
+
+#define NVME_ALL_NAMESPACES  0xFFFFFFFF
+
+// MU_CHANGE [BEGIN] - Support alternative hardware queue sizes in NVME driver
```
Comment (makubacki) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h]: This change came from a different commit https://github.com/microsoft/mu_basecore/commit/9837db39da115a1fa4a322d13adafbba862212d8.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c (hunk 1 above)
Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c]: Should be restored in latest. Thanks for catching it.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h (hunk 2 above)
Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h]: Should be fixed in latest, thanks for catching it.

File: MdeModulePkg/Test/MdeModulePkgHostTest.dsc
Comment (makubacki) [inline MdeModulePkg/Test/MdeModulePkgHostTest.dsc:1]: It looks like there's a lot of changes coming into this file ([MdeModulePkg/Test/MdeModulePkgHostTest.dsc](https://github.com/microsoft/mu_basecore/pull/904/files#diff-272848d461635a235f3062662f8088f06edb1d9c6b868128f3b48cf70ec5ef82)) that shouldn't be

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h [hunk 3]
```diff
@@ -2,8 +2,9 @@
   NvmExpressDxe driver is used to manage non-volatile memory subsystem which follows
   NVM Express specification.
 
-  (C) Copyright 2016 Hewlett Packard Enterprise Development LP<BR>
-  Copyright (c) 2013 - 2019, Intel Corporation. All rights reserved.<BR>
+  (C) Copyright 2016 Hewlett Packard Enterprise Development LP
+  Copyright (c) 2013 - 2019, Intel Corporation. All rights reserved.
```
Comment (os-d) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h]: We don't need to touch the other copyrights.

File: MdeModulePkg/Include/Protocol/MediaSanitize.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
+/** @file
+  This file defines the Media Sanitize Protocol.
+
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef MEDIA_SANITIZE_PROTOCOL_H__
```
Comment (os-d) [inline MdeModulePkg/Include/Protocol/MediaSanitize.h]: nit: remove second underscore at end

File: MdeModulePkg/Test/MdeModulePkgHostTest.dsc [hunk 5]
```diff
@@ -1,66 +1,71 @@
...
+# MdeModulePkg DSC file used to build host-based unit tests.
+#
+# Copyright (c) 2019 - 2020, Intel Corporation. All rights reserved.<BR>
+# Copyright (C) Microsoft Corporation.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+##
+
+[Defines]
+  PLATFORM_NAME           = MdeModulePkgHostTest
```
Comment (os-d) [inline MdeModulePkg/Test/MdeModulePkgHostTest.dsc]: Line endings?

**Review by os-d — APPROVED** [review]

File: MdeModulePkg/Test/MdeModulePkgHostTest.dsc (hunk 5 above)
Comment (apop5) [reply MdeModulePkg/Test/MdeModulePkgHostTest.dsc]: Fixed, thanks for seeing it.

**Review by makubacki — APPROVED** [review]

**Pull request merged**